File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.js
function solution(s) {
    let minLength = s.length;  // 압축 안 된 경우 (최대값)
    
    // 1부터 s.length/2까지 모든 단위 길이 시도
    for (let unit = 1; unit <= Math.floor(s.length / 2); unit++) {
        let compressed = "";
        let count = 1;
        let prev = s.slice(0, unit);  // 첫 번째 패턴
        
        // unit 길이씩 잘라가며 비교
        for (let i = unit; i < s.length; i += unit) {
            const current = s.slice(i, i + unit);
            
            if (prev === current) {
                count++;
            } else {
                // 이전 패턴 처리
                compressed += (count > 1 ? count : "") + prev;
                prev = current;
                count = 1;
            }
        }
        
        // 마지막 패턴 처리
        compressed += (count > 1 ? count : "") + prev;
        
        minLength = Math.min(minLength, compressed.length);
    }
    
    return minLength;
}

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.py
def solution(s):
    n = len(s)
    compression_length_array = []  # 1~len까지 압축했을때 길이 값
    if n == 1:
        return 1
    for split_size in range(1, n // 2 + 1):
        compressed = ""
        # string 갯수 단위로 쪼개기 *
        splited = [
            s[i:i + split_size] for i in range(0, n, split_size)
        ]
        count = 1

        for j in range(0, len(splited)-1):
            cur, next = splited[j], splited[j + 1]
            if cur == next:
                count += 1
            else:  # 이전 문자와 다르다면
                if count > 1:
                    compressed += (str(count) + cur)
                else:  # 문자가 반복되지 않아 한번만 나타난 경우 1은 생략함
                    compressed += cur
                count = 1  # 초기화
        if count > 1:
            compressed += (str(count) + splited[-1])
        else:  # 문자가 반복되지 않아 한번만 나타난 경우 1은 생략함
            compressed += splited[-1]
        compression_length_array.append(len(compressed))

    return min(compression_length_array)  # 최솟값 리턴

